test/algorithms/sort/bucket-sort.test.ts

```diff
@@ -0,0 +1,46 @@
+import bucketSort from '../../../src/algorithms/sort/bucket-sort'
+import testCases from './test-cases'
+
+describe('bucketSort', () => {
+  console.error = jest.fn()
+
+  beforeEach(() => {
+    jest.clearAllMocks()
+  })
+
+  for (const [description, inputArray] of Object.entries(testCases)) {
+    it(`correctly sorts ${description}`, () => {
+      const soln = [...inputArray]
+      soln.sort((a, b) => a - b) // sort the copy inplace using native JS function
+      bucketSort(inputArray) //sort the original inplace using bucketSort
+      expect(inputArray).toEqual(soln)
+    })
+  }
+
+  it('correctly sorts when only posBuckets is specified', () => {
+    const array = [5, 3, 4, 1, 2]
+    const soln = [...array].sort((a, b) => a - b)
+    bucketSort(array, 1)
+    expect(array).toEqual(soln)
+  })
+
+  it('correctly sorts when only negBuckets is specified', () => {
+    const array = [-5, -3, -4, -1, -2]
+    const soln = [...array].sort((a, b) => a - b)
+    bucketSort(array, undefined, 1)
+    expect(array).toEqual(soln)
+  })
+
+  it('correctly sorts when both posBuckets and negBuckets are specified', () => {
+    const array = [-5, -3, 7, -4, -1, -2, 9, 0]
+    const soln = [...array].sort((a, b) => a - b)
+    bucketSort(array, 5, 3)
```

Comment:
wait, how is this test passing? shouldn't it be `bucketSort(array, 3, 5)`?

Reply:
The number of buckets specified shouldn't affect the algorithm's ability to sort correctly as long as there is at least one bucket to place elements into. 

If there are fewer buckets than array elements, some buckets will end up with more than one element. In this case, I think the algorithm's runtime would approach O(n^2) as the ratio of element : bucket approaches the length of the array we're sorting (our worst case time complexity). If there are more buckets than elements, some buckets will end up empty. In this second case, the run time would be O(n + k), it's just that k (the number of buckets) isn't optimal.

Since I give users the option to specify the optimal number of buckets, I think we should make sure that it still sorts correctly if they choose too many or too few buckets.